File: src/cli/discover.ts
/**
 * File-system discovery of test files.
 *
 * Design goals:
 *   - Walks a root directory
 *   - Honours `.gitignore` (basic pattern subset: basename match, leading-`/`
 *     root anchor, directory-trailing `/`, simple `*` glob, path-scoped
 *     `a/*.ts` patterns)
 *   - Uses `lstat` so symlinks are NOT followed (cycle-safe, matches Rust's
 *     `ignore::WalkBuilder::follow_links(false)`)
 *   - Tracks visited absolute paths so any escape via junctions is capped
 *   - Emits a warning on permission-denied subtrees (so silent tests-gone-missing
 *     is visible) while still returning the discoverable set
 */

import { existsSync, readFileSync } from "node:fs";
import { lstat, readdir } from "node:fs/promises";
import path from "node:path";

export interface DiscoveryOptions {
	/** Filename suffixes that mark a test file (e.g. `.test.ts`). */
	suffixes?: string[];
	/** Directory basenames pruned from the walk. */
	hardExcludes?: string[];
	/** Read `.gitignore` at `root` + every descendant and apply rules. */
	honourGitignore?: boolean;
	/** Called with a human-readable message when a directory is skipped
	 *  because of an IO error (ENOENT, EACCES). Default: `console.warn`. */
	onWarn?: (message: string) => void;
}

const DEFAULT_SUFFIXES = [
	".test.ts",
	".test.tsx",
	".test.js",
	".test.mjs",
	".test.cjs",
	".spec.ts",
	".spec.tsx",
	".spec.js",
	".spec.mjs",
	".spec.cjs",
];

const DEFAULT_HARD_EXCLUDES = [
	"node_modules",
	"dist",
	"build",
	"coverage",
	".git",
	".wolf",
	"target",
	".next",
];

interface GitignorePattern {
	raw: string;
	/** If true, pattern is anchored to the directory that defined it. */
	anchored: boolean;
	/** If true, pattern only matches directories (trailing `/`). */
	dirOnly: boolean;
	/** Regex compiled from the literal pattern, matched against a relative path. */
	regex: RegExp;
}

function compilePattern(line: string): GitignorePattern | undefined {
	let p = line.trim();
	if (!p || p.startsWith("#")) return undefined;
	const anchored = p.startsWith("/");
	if (anchored) p = p.slice(1);
	const dirOnly = p.endsWith("/");
	if (dirOnly) p = p.slice(0, -1);
	if (!p) return undefined;
	// Translate a tiny glob subset to regex:
	//   `*` → `[^/]*`
	//   `**` → `.*`
	const escaped = p
		.replace(/[.+^${}()|[\]\\]/g, "\\$&")
		.replace(/\*\*/g, "__DOUBLESTAR__")
		.replace(/\*/g, "[^/]*")
		.replace(/__DOUBLESTAR__/g, ".*");
	const regex = anchored
		? new RegExp(`^${escaped}(?:/.*)?$`)
		: new RegExp(`(?:^|/)${escaped}(?:/.*)?$`);
	return { raw: line, anchored, dirOnly, regex };
}

function readGitignore(dir: string): GitignorePattern[] {
	const file = path.join(dir, ".gitignore");
	if (!existsSync(file)) return [];
	try {
		return readFileSync(file, "utf8")
			.split("\n")
			.map(compilePattern)
			.filter((p): p is GitignorePattern => p !== undefined);
	} catch {
		return [];
	}
}

function matches(
	pattern: GitignorePattern,
	relPath: string,
	isDir: boolean,
): boolean {
	if (pattern.dirOnly && !isDir) return false;
	return pattern.regex.test(relPath);
}

export async function discover(
	root: string,
	options: DiscoveryOptions = {},
): Promise<string[]> {
	const suffixes = options.suffixes ?? DEFAULT_SUFFIXES;
	const hardExcludes = new Set(options.hardExcludes ?? DEFAULT_HARD_EXCLUDES);
	const honourGitignore = options.honourGitignore ?? true;
	const warn = options.onWarn ?? ((m) => process.stderr.write(`helix: ${m}\n`));

	const results: string[] = [];
	const absRoot = path.isAbsolute(root) ? root : path.resolve(root);
	const visited = new Set<string>();

	async function walk(
		dir: string,
		relativeToRoot: string,
		inherited: GitignorePattern[],
	): Promise<void> {
		const realDir = path.resolve(dir);
		if (visited.has(realDir)) return;
		visited.add(realDir);

		let entries: string[];
		try {
			entries = await readdir(dir);
		} catch (err) {
			warn(`skipping ${dir}: ${(err as NodeJS.ErrnoException).code ?? err}`);
			return;
		}
		const local = honourGitignore
			? [...inherited, ...readGitignore(dir)]
			: inherited;

		for (const name of entries) {
			if (hardExcludes.has(name)) continue;
			const relPath = relativeToRoot ? `${relativeToRoot}/${name}` : name;
			const fullPath = path.join(dir, name);
			let st: Awaited<ReturnType<typeof lstat>>;
			try {
				st = await lstat(fullPath);
			} catch (err) {
				warn(
					`skipping ${fullPath}: ${(err as NodeJS.ErrnoException).code ?? err}`,
				);
				continue;
			}
			// Symlinks are NOT followed. Users who need symlink-following can
			// walk the target manually; matches Rust's `follow_links(false)`.
			if (st.isSymbolicLink()) continue;
			if (local.some((p) => matches(p, relPath, st.isDirectory()))) continue;
			if (st.isDirectory()) {
				await walk(fullPath, relPath, local);
				continue;
			}
			if (!st.isFile()) continue;
			if (suffixes.some((s) => name.endsWith(s))) {
				results.push(fullPath);
			}
		}
	}

	await walk(absRoot, "", []);
	results.sort();
	return results;
}

File: crates/ream-test-core/src/discover.rs
//! Test-file discovery. Walks a root directory respecting `.gitignore` /
//! `.ignore` (via the `ignore` crate) and matches filenames against
//! include/exclude glob patterns.
//!
//! Defaults are tuned for the JS/TS ecosystem: `*.test.{ts,tsx,js,mjs,cjs}`
//! and `*.spec.{ts,tsx,js,mjs,cjs}` under the root. `node_modules/` and
//! build outputs are excluded upfront.

use ignore::WalkBuilder;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone)]
pub struct DiscoveryOptions {
    /// Filename suffixes that mark a test file. Matched case-sensitively
    /// against the file's `.file_name()`. E.g. `.test.ts`.
    pub suffixes: Vec<String>,
    /// Directory names (relative to root) to prune from the walk. Applied
    /// in addition to `.gitignore` semantics.
    pub hard_excludes: Vec<String>,
    /// If true, also descend into `node_modules/` (default: false).
    pub include_node_modules: bool,
}

impl Default for DiscoveryOptions {
    fn default() -> Self {
        Self {
            suffixes: vec![
                ".test.ts".into(),
                ".test.tsx".into(),
                ".test.js".into(),
                ".test.mjs".into(),
                ".test.cjs".into(),
                ".spec.ts".into(),
                ".spec.tsx".into(),
                ".spec.js".into(),
                ".spec.mjs".into(),
                ".spec.cjs".into(),
            ],
            hard_excludes: vec![
                "node_modules".into(),
                "dist".into(),
                "build".into(),
                "coverage".into(),
                ".git".into(),
                ".wolf".into(),
                "target".into(),
            ],
            include_node_modules: false,
        }
    }
}

/// Walk `root` and return every test file, sorted lexicographically.
///
/// The walker uses `ignore::WalkBuilder` which honours `.gitignore`, `.ignore`,
/// and global ignore files. Hidden files (dotfiles) are also skipped.
pub fn discover(root: &Path, opts: &DiscoveryOptions) -> std::io::Result<Vec<PathBuf>> {
    let mut builder = WalkBuilder::new(root);
    builder
        .hidden(true)
        .git_ignore(true)
        .git_exclude(true)
        .follow_links(false)
        .require_git(false);

    // Filter out hard-excluded directories early so we don't even descend.
    let excludes: Vec<String> = opts.hard_excludes.clone();
    let allow_node_modules = opts.include_node_modules;
    builder.filter_entry(move |entry| {
        let name = match entry.file_name().to_str() {
            Some(n) => n,
            None => return true,
        };
        if !allow_node_modules && name == "node_modules" {
            return false;
        }
        !excludes.iter().any(|ex| ex == name)
    });

    let mut out: Vec<PathBuf> = Vec::new();
    for entry in builder.build() {
        let entry = match entry {
            Ok(e) => e,
            Err(_) => continue,
        };
        let file_type = match entry.file_type() {
            Some(t) => t,
            None => continue,
        };
        if !file_type.is_file() {
            continue;
        }
        let name = match entry.file_name().to_str() {
            Some(n) => n,
            None => continue,
        };
        if opts.suffixes.iter().any(|s| name.ends_with(s.as_str())) {
            out.push(entry.into_path());
        }
    }
    out.sort();
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::io::Write;

    fn touch(path: &Path) {
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent).unwrap();
        }
        let mut f = fs::File::create(path).unwrap();
        writeln!(f, "// test fixture").unwrap();
    }

    #[test]
    fn finds_test_files_and_ignores_node_modules() {
        let tmp = tempfile_dir();
        touch(&tmp.join("a/foo.test.ts"));
        touch(&tmp.join("a/bar.spec.ts"));
        touch(&tmp.join("a/not-a-test.ts"));
        touch(&tmp.join("node_modules/x/baz.test.ts"));
        touch(&tmp.join("dist/built.test.ts"));

        let opts = DiscoveryOptions::default();
        let files = discover(&tmp, &opts).unwrap();
        let names: Vec<String> = files
            .iter()
            .map(|p| p.file_name().unwrap().to_string_lossy().into_owned())
            .collect();
        assert!(names.contains(&"foo.test.ts".to_string()));
        assert!(names.contains(&"bar.spec.ts".to_string()));
        assert!(!names.contains(&"not-a-test.ts".to_string()));
        assert!(!names.contains(&"baz.test.ts".to_string()));
        assert!(!names.contains(&"built.test.ts".to_string()));
    }

    #[test]
    fn results_are_sorted() {
        let tmp = tempfile_dir();
        touch(&tmp.join("z.test.ts"));
        touch(&tmp.join("a.test.ts"));
        touch(&tmp.join("m.test.ts"));
        let files = discover(&tmp, &DiscoveryOptions::default()).unwrap();
        assert!(files.len() == 3);
        assert!(files[0].ends_with("a.test.ts"));
        assert!(files[1].ends_with("m.test.ts"));
        assert!(files[2].ends_with("z.test.ts"));
    }

    fn tempfile_dir() -> PathBuf {
        let mut dir = std::env::temp_dir();
        let id = format!(
            "helix-test-{}-{}",
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_nanos()
        );
        dir.push(id);
        fs::create_dir_all(&dir).unwrap();
        dir
    }
}
